evals/solvers/providers/openai/whisper_solver.py

```
@@ -3,17 +3,74 @@
 import io
 from concurrent.futures import ProcessPoolExecutor
 from typing import Any, Optional
-
 import librosa
 import openai
 import torch
 import torch.distributed
 import torch.multiprocessing as mp
 import transformers
-
 from evals.solvers.solver import NestedSolver, Solver, SolverResult, SolverSpec
 from evals.task_state import Message, TaskState
 
+class TranscriptionSolver(Solver):
+    def __init__(
+            self,
+            base_url: Optional[str] = None,
+            api_key: Optional[str] = None,
+            postprocessors: list[str] = [],
+            registry: Any = None,
+    ):
+        super().__init__(postprocessors=postprocessors, registry=registry)
+        self.base_url = base_url
+        self.api_key = api_key
+        self.client: Optional[openai.OpenAI] = None
+
+    def _solve(self, task_state: TaskState, **kwargs) -> SolverResult:
+        return SolverResult(self._process_msgs(task_state.messages))
+    
+    def _process_msgs(self, raw_msgs: list[Message]) -> list[Message]:
+        # process the first message with an audio url
+        for msg in raw_msgs:
+            if msg.role == "user" and isinstance(msg.content, list):
+                for part in msg.content:
+                    if part["type"] == "audio_url":
+                        url = part["audio_url"]["url"]
+                        if url.startswith("data:audio/x-wav;base64,"):
+                            wav_b64 = url.split(",")[1]
+                            wav_bytes = base64.b64decode(wav_b64)
+                            text = self._transcribe(wav_bytes)
+                            return text
+        return ""    
```

If there is no audio wouldn't this be confounded with silent audio chunks where the ASR produces an empty transcript? Is there another way to indicate that there is no audio chunk so its not included in the evaluation?

---

Every transcription sample should have an audio chunk right? I see your point that we might somehow want to flag those samples without chunks to differentiate them from samples with chunks that also come back with `""`. But would this ever happen?

---

True, we could also throw an exception here if there isn't an audio chunk right. Other solvers in this repo don't do this though so I'm fine leaving it the way it is, and patching it if we do eventually add an eval that has missing audio for some reason

---

I added an exception. I think it would be worse for it to silently fail